### PR TITLE
Add tests and impl for Pages class and associated iterator

### PR DESCRIPTION
**Pull request by software-dov**
The Pages class is the next level up in the PaginatedResult
abstraction hierarchy above PageResult. It provides:
* a single constructor that takes a (copied) closure that retrieves
  successive pages from the API service.
* iterator-returning begin() and end() methods.

The refactor for PageResult includes turning it into a nested class in
PageResult and removing all ElementAccessor members and parameters:
ElementAccessor is intended to be a lightweight, stateless, uniform wrapper, and is
now used purely as a default-constructed throwaway.

### CODE REVIEW

**Comment by michaelbausor** [comment]
What is the benefit of making PageResult nested vs a standalone class?

**Comment by software-dov** [comment]
The nested class obviates duplication of the templates and template constraints and means there aren't as many forward declarations necessary. It's purely a maintenance and convenience decision.

**Comment by software-dov** [comment]
Open review bump
@vam-google @michaelbausor @coryan

File: gax/pagination.h [hunk 1]
```diff
@@ -16,60 +16,145 @@
...
     }
     bool operator!=(iterator const& rhs) const { return !(*this == rhs); }
 
-    iterator(FieldIterator current) : current_(std::move(current)) {}
+   private:
+    friend Pages;
+    // Note: don't need to provide a const-lvalue receiving version of the
+    // constructor since iterators always take ownership of a page and pages can
+    // be very expensive to copy if they have a lot of repeated elements.
+    iterator(PageType&& page_result, NextPageRetriever get_next_page,
```
Comment (coryan) [inline gax/pagination.h]: Consider `PageType page_result` and make the caller `std::move()`.

File: gax/pagination.h [hunk 2]
```diff
@@ -16,60 +16,145 @@
...
-  // access), we don't have to define all the variants for references.
-  // We just want to take ownership of a page.
-  PageResult(PageType&& raw_page, ElementAccessor accessor)
-      : raw_page_(std::move(raw_page)), accessor_(accessor) {}
 
-  iterator begin() { return accessor_(raw_page_)->begin(); }
-  iterator end() { return accessor_(raw_page_)->end(); }
+  // Note that a cap of 0 indicates that an iterator will only terminate
+  // when the next page token is empty.
+  Pages(NextPageRetriever get_next_page, int cap)
```
Comment (coryan) [inline gax/pagination.h]: What is `cap`? Is that a limit on the number of pages or a limit on the number of results?  I can figure it out, but make it self-describing or document what it is.

**Comment by michaelbausor** [comment]
I'm not convinced that a nested class is the way to go. Is the benefit for maintenance significant? I think that the `PageResult` class should logically be considered separate from a `Pages` class. It is "lower" in the hierarchy of classes being implemented for pagination, and could be returned directly from an RPC method on the class, letting users interact with it without reference to a `Pages` class.

File: gax/pagination.h [hunk 3]
```diff
@@ -16,60 +16,145 @@
...
 namespace google {
 namespace gax {
 
-template <typename ElementType, typename PageType, typename ElementAccessor>
-class PageResult {
-  using FieldIterator =
-      typename std::remove_pointer<typename gax::internal::invoke_result_t<
-          ElementAccessor, PageType&>>::type::iterator;
-
+template <
```
Comment (coryan) [inline gax/pagination.h]: Do you folks create class comments?

File: gax/pagination.h (hunk 1 above)
Comment (software-dov) [reply gax/pagination.h]: Just for my own peace of mind I'll do a quick benchmark first to see if the extra move construction has a noticeable impact. Will update with the results afterwards.

File: gax/pagination.h (hunk 2 above)
Comment (software-dov) [reply gax/pagination.h]: It's a cap on the number of pages to fetch. I'll update the name and annotate it in method comments.

**Comment by software-dov** [comment]
I'd like to not duplicate the template boilerplate if at all possible, and I'd like to keep the accessor and next-page retrieval in the template. I'll try some cleverness and report back with how onerous the predicted maintenance burden becomes.

File: gax/pagination.h (hunk 3 above)
Comment (software-dov) [reply gax/pagination.h]: I'm not regular about it :P Will add.

**Comment by software-dov** [comment]
> I'd like to not duplicate the template boilerplate if at all possible, and I'd like to keep the accessor and next-page retrieval in the template. I'll try some cleverness and report back with how onerous the predicted maintenance burden becomes.

Result is not as bad as I had feared. There's some duplication of the template, but it should be managable.

File: gax/pagination.h (hunk 1 above)
Comment (software-dov) [reply gax/pagination.h]: In the benchmark, for a repeated field with ~50 elements, having an rvalue ref parameter is about 11% faster. With 100 elements the speedup shrinks to ~1.7%, and with 200 elements it shrinks to less than a percent. Going with the value parameter.

**Comment by software-dov** [comment]
Open review bump
@vam-google @michaelbausor @coryan

File: gax/pagination.h [hunk 4]
```diff
@@ -51,25 +74,185 @@ class PageResult {
...
-  // We just want to take ownership of a page.
-  PageResult(PageType&& raw_page, ElementAccessor accessor)
-      : raw_page_(std::move(raw_page)), accessor_(accessor) {}
 
-  iterator begin() { return accessor_(raw_page_)->begin(); }
-  iterator end() { return accessor_(raw_page_)->end(); }
+  PageResult(PageType const& raw_page) : raw_page_(raw_page) {}
+  PageResult(PageType&& raw_page) : raw_page_(std::move(raw_page)) {}
 
+  iterator begin() { return ElementAccessor{}(raw_page_)->begin(); }
```
Comment (michaelbausor) [inline gax/pagination.h]: Why construct the ElementAccessor over and over again? Can't we retain the accessor_ field and use that? Or does that add to the memory footprint and not actually result in any improvement?

File: gax/pagination.h [hunk 5]
```diff
@@ -51,25 +74,185 @@ class PageResult {
...
+   *
+   * @retun the token for the next page in the sequence
+   * or the empty string if this is the last page.
+   */
   std::string NextPageToken() const { return raw_page_.next_page_token(); }
+
+  /**
+   * @brief Get the underlying page message.
+   *
+   * This can be useful if the page has custom metadata fields.
```
Comment (michaelbausor) [inline gax/pagination.h]: metadata fields? Should this say "fields other than the repeated field or page token?"

File: gax/pagination.h [hunk 6]
```diff
@@ -51,25 +74,185 @@ class PageResult {
...
+    using value_type = PageResultT;
+    using difference_type = std::ptrdiff_t;
+    using pointer = PageResultT*;
+    using reference = PageResultT&;
+
+    PageResultT const& operator*() const { return page_result_; }
+    PageResultT const* operator->() const { return &page_result_; }
+    iterator& operator++() {
+      // Note: if the rpc fails, the page will be untouched,
+      // i.e. will be have an empty page token and element collection.
```
Comment (michaelbausor) [inline gax/pagination.h]: s/will be have/will have

File: gax/pagination.h (hunk 4 above)
Comment (software-dov) [reply gax/pagination.h]: The accessor has no fields and a trivial constructor, and the overload to operator() should hopefully be small enough to be inlined.
Saving an instance of ElementAccessor requires at least 1 byte of space; I haven't run any benchmarks, but it should never be more expensive to recreate the ElementAccessor whenever it's used.

File: gax/pagination.h (hunk 5 above)
Comment (software-dov) [reply gax/pagination.h]: Will do.

File: gax/pagination.h (hunk 6 above)
Comment (software-dov) [reply gax/pagination.h]: Will do.

**Pull request merged**